Crash (Error) when attempting to save video frame on Pixel 3 emulator with API 29

First observed and reported by @jd-alexander in #297, managed to reproduce in `develop`, this is the stacktrace:

```
2020-04-10 13:35:11.616 14743-15252/com.automattic.portkey D/SurfaceUtils: disconnecting from surface 0xbdd71008, reason disconnectFromSurface
2020-04-10 13:35:11.627 14743-15233/com.automattic.portkey E/AndroidRuntime: FATAL EXCEPTION: pool-3-thread-1
    Process: com.automattic.portkey, PID: 14743
    java.lang.Error: Could not shutdown mediaExtractor, codecs and mediaMuxer pipeline.
        at com.daasuu.mp4compose.composer.Mp4ComposerEngine.compose(Mp4ComposerEngine.kt:240)
        at com.daasuu.mp4compose.composer.Mp4ComposerEngine.composeFromVideoSource(Mp4ComposerEngine.kt:69)
        at com.daasuu.mp4compose.composer.Mp4Composer$start$1.run(Mp4Composer.kt:191)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: java.lang.IllegalStateException
        at android.media.MediaCodec.native_stop(Native Method)
        at android.media.MediaCodec.stop(MediaCodec.java:2147)
        at com.daasuu.mp4compose.composer.VideoComposer.release(VideoComposer.kt:214)
        at com.daasuu.mp4compose.composer.Mp4ComposerEngine.compose(Mp4ComposerEngine.kt:226)
        at com.daasuu.mp4compose.composer.Mp4ComposerEngine.composeFromVideoSource(Mp4ComposerEngine.kt:69) 
        at com.daasuu.mp4compose.composer.Mp4Composer$start$1.run(Mp4Composer.kt:191) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
2020-04-10 13:35:11.659 14743-15233/com.automattic.portkey D/io.sentry.android.event.helper.AndroidEventBuilderHelper: Proguard UUIDs file not found.
2020-04-10 13:35:11.700 1775-15259/? E/ResolverController: No valid NAT64 prefix (101, <unspecified>/0)

```

#### Steps to reproduce
1. open the app on the Pixel 3 API29 emu, tap on the + FAB to get to capture mod
2. record a video by tapping the capture button and holding for a few seconds
3. add some text or emoji
4. tap NEXT to save
5. 💥 

#### Research

I investigated it a bit, and found the following:

The app was closing because the `mp4composer` module throws an `Error`  in [this line](https://github.com/Automattic/portkey-android/blob/develop/mp4compose/src/main/java/com/daasuu/mp4compose/composer/Mp4ComposerEngine.kt#L240).
This is actually after having interrupted execution and calling `release()` on videoComposer, audioComposer, and mediaExtractor.
**tl;dr**  we're passing a target video size that matches the device screen canvas size given we're aiming at providing a WYSYWIG kind of experience, as can be seen in this commit here https://github.com/Automattic/portkey-android/commit/b25545e1873022528e8565b89880ebe80b186b17#diff-1dfe56dc06fa5c09f0d3ac32efea40d9R841-R845 

The problem wasn't happening for instance on a Pixel 2 emulator running Android 8.1 (and FWIW it wasn't until now that the problem surfaced), and hence it makes sense because the screen sizes and video codec capabilities vary from phone to phone. It was just the case that the Pixel 2 has a screen size for which a video codec has a matching format, and thus it never crashed.

Just testing by hand, I figured out the values for size on the Pixel 3 where:
- 1440x2792 --> this is the canvas size (in portrait)
- 1440x2560 --> this is the closest compatible video size ([WQHD](https://en.wikipedia.org/wiki/Graphics_display_resolution)).

When using the second value, everything goes fine.

The problem that this poses is that we'll be always missing part of the screen when converting video frames (for example, if a user placed some emoji in a corner of the screen), but it shouldn't matter in practice given it's a small portion and, coincidentally, users won't have all the space available to drag and drop stuff given we're using part of the screen for controls (see https://github.com/Automattic/portkey-android/pull/295).

Ideally this issue should be solved by trying to discover the closest supported video size for encoding video in runtime, we're exploring our options there.

--- a/mp4compose/src/main/java/com/daasuu/mp4compose/composer/Mp4Composer.kt
+++ b/mp4compose/src/main/java/com/daasuu/mp4compose/composer/Mp4Composer.kt
@@ -71,6 +71,11 @@ class Mp4Composer {
         return this
     }
 
+    fun size(size: Size): Mp4Composer {
+        this.outputResolution = size
+        return this
+    }
+
     fun videoBitrate(bitrate: Int): Mp4Composer {
         this.bitrate = bitrate
         return this

--- a/photoeditor/src/main/java/com/automattic/photoeditor/PhotoEditor.kt
+++ b/photoeditor/src/main/java/com/automattic/photoeditor/PhotoEditor.kt
@@ -10,6 +10,7 @@ import android.media.MediaMetadataRetriever
 import android.net.Uri
 import android.text.TextUtils
 import android.util.Log
+import android.util.Size
 import android.util.TypedValue
 import android.view.Gravity
 import android.view.LayoutInflater
@@ -710,7 +711,9 @@ class PhotoEditor private constructor(builder: Builder) :
             // IMPORTANT: as we aim at a WYSIWYG UX, we need to produce a video of size equal to that of the phone
             // screen, given the user may be seeing a letterbox landscape video and placing emoji / text around
             // the black parts of the screen.
-            .size(originalCanvasWidth, originalCanvasHeight)
+            // .size(originalCanvasWidth, originalCanvasHeight)
+            // normalize output video size if actual screen size does not match a "normal" video size
+            .size(normalizeTargetVideoSize(originalCanvasWidth, originalCanvasHeight))
             .fillMode(FillMode.PRESERVE_ASPECT_FIT)
             .filter(GlFilterGroup(filterCollection))
             .listener(object : Mp4Composer.Listener {
@@ -735,6 +738,54 @@ class PhotoEditor private constructor(builder: Builder) :
                 }
             })
             .start()
+    }
+
+    private fun normalizeTargetVideoSize(
+        requestedWidth: Int,
+        requestedHeight: Int
+    ): Size {
+        var adjustedSize = Size(requestedWidth, requestedHeight)
+        // As per CDD, all android devices running API level 21 (our minSdk) with H.264 codec must support 720 x 480 px.
+        // see https://source.android.com/compatibility/5.0/android-5.0-cdd#5_2_video_encoding
+        // MUST
+        // - 320 x 240 px
+        // - 720 x 480 px
+        // SHOULD (when hardware available)
+        // - 1280 x 720 px
+        // - 1920 x 1080 px
+
+        // the other formats (2160p, 1440p, 1080p etc) are "popular" ones found on many devices
+        // note we reverse the width/height given we support portrait mode only.
+        when {
+            // 2160p = 3840x2160
+            (requestedWidth % 2160 == 0) -> {
+                adjustedSize = Size(2160, 3840)
+            }
+
+            // 1440p = 2560x1440
+            (requestedWidth % 1440 == 0) -> {
+                adjustedSize = Size(1440, 2560)
+            }
+
+            // 1080p = 1920x1080
+            (requestedWidth % 1080 == 0) -> {
+                adjustedSize = Size(1080, 1920)
+            }
+
+            // 720p = 1280x720
+            (requestedWidth % 720 == 0) -> {
+                adjustedSize = Size(720, 1280)
+            }
+
+            (requestedWidth % 480 == 0) -> {
+                adjustedSize = Size(480, 720)
+            }
+
+            (requestedWidth % 240 == 0) -> {
+                adjustedSize = Size(240, 320)
+            }
+        }
+        return adjustedSize
     }
 
     /**

